### PR TITLE
Manage libseccomp-sys by hand

### PR DESCRIPTION
**Pull request by rusty-snake**
Todos:
- [x] Test static linking (CI)
- [x] Check other non-code files and updated them if necessary
- [x] repr(C) on enums can cause UB because we never really know the size of the used integer (for example if libseccomp was compile with `-fshort-enums`). However it looks like there is less to nothing that can be done about this (on our end).
- [x] ~How does SCMP_CMP32 work? Can we add bindings for it? (nice to have)~
- [x] Add doc comments
- [x] ~Provided `seccomp-syscalls.h.` and `SCMP_SYS` (i.e. compile time resolution of syscalls).~ goes into it's own issue
- [x] keep tool, remove tool, add tool.sh?

Checklist:
- [x] all enums/struct have repr(C)
- [x] all constants/enums have the right value (copy&paste mistakes, ...)
- [x] all structfiles have the right type
- [x] all constants have the right type (`#define` does not have a type)
- [x] `FIXME: *const c_void`

### CODE REVIEW

**Comment by ManaSugi** [comment]
@rusty-snake Thank you for creating the PR!
Did you create `libseccomp-sys` by using the existing `tool`?
I think we do not have to maintain `SCMP_` prefix function, `AUDIT_`  variable and so so. Hence, `tool` creates `libseccomp-sys` without them. In addition, we need `tool` to update `libseccomp-sys` for new version, so It's worth leaving `tool` for the developer.

File: libseccomp-sys/src/lib.rs [hunk 1]
```diff
@@ -1,24 +1,381 @@
...
+const AUDIT_ARCH_PARISC: u32 = EM_PARISC;
+const AUDIT_ARCH_PARISC64: u32 = EM_PARISC | __AUDIT_ARCH_64BIT;
+const AUDIT_ARCH_PPC: u32 = EM_PPC;
+const AUDIT_ARCH_PPC64: u32 = EM_PPC64 | __AUDIT_ARCH_64BIT;
+const AUDIT_ARCH_PPC64LE: u32 = EM_PPC64 | __AUDIT_ARCH_64BIT | __AUDIT_ARCH_LE;
+const AUDIT_ARCH_RISCV64: u32 = EM_RISCV | __AUDIT_ARCH_64BIT | __AUDIT_ARCH_LE;
+const AUDIT_ARCH_S390: u32 = EM_S390;
+const AUDIT_ARCH_S390X: u32 = EM_S390 | __AUDIT_ARCH_64BIT;
+const AUDIT_ARCH_X86_64: u32 = EM_X86_64 | __AUDIT_ARCH_64BIT | __AUDIT_ARCH_LE;
+// ]]
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:94]: We do not need these.

File: libseccomp-sys/src/lib.rs [hunk 2]
```diff
@@ -1,24 +1,381 @@
-// SPDX-License-Identifier: Apache-2.0 or MIT
-//
-// Copyright 2021 Sony Group Corporation
-//
-
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:5]: Could you leave the SPDX license and copyright?

File: libseccomp-sys/src/lib.rs [hunk 3]
```diff
@@ -1,24 +1,381 @@
...
-include!("./libseccomp_bindings.rs");
+pub const fn SCMP_A5(
+    op: scmp_compare,
+    datum_a: scmp_datum_t,
+    datum_b: scmp_datum_t,
+) -> scmp_arg_cmp {
+    SCMP_A5_65(op, datum_a, datum_b)
+}
+
+// SCMP_A5_32
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:264]: We do not need these.

**Comment by rusty-snake** [comment]
>  Did you create libseccomp-sys by using the existing tool?

Actually, I just went through `seccomp.h`.

File: libseccomp-sys/src/lib.rs [hunk 4]
```diff
@@ -3,22 +3,196 @@
...
+pub const SCMP_ARCH_MIPSEL64N32: u32 = 0xe0000008;
+pub const SCMP_ARCH_PPC: u32 = 0x14;
+pub const SCMP_ARCH_PPC64: u32 = 0x80000015;
+pub const SCMP_ARCH_PPC64LE: u32 = 0xc0000015;
+pub const SCMP_ARCH_S390: u32 = 0x16;
+pub const SCMP_ARCH_S390X: u32 = 0x80000016;
+pub const SCMP_ARCH_PARISC: u32 = 0xf;
+pub const SCMP_ARCH_PARISC64: u32 = 0x8000000f;
+pub const SCMP_ARCH_RISCV64: u32 = 0xc00000f3;
+
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:82]: Please add:
```rust
pub const SCMP_ACT_MASK: u32 = 0xFFFF0000;
```
It will be used by some functions for the future.

File: libseccomp-sys/src/lib.rs [hunk 5]
```diff
@@ -3,22 +3,196 @@
...
+pub const SCMP_ACT_KILL_PROCESS: u32 = 0x80000000;
+pub const SCMP_ACT_KILL_THREAD: u32 = 0x00000000;
+pub const SCMP_ACT_KILL: u32 = SCMP_ACT_KILL_THREAD;
+pub const SCMP_ACT_TRAP: u32 = 0x00030000;
+pub const SCMP_ACT_NOTIFY: u32 = 0x7fc00000;
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:87]: Please add:
```rust
pub const SCMP_ACT_ERRNO_MASK: u32 = 0x00050000;
pub const SCMP_ACT_TRACE_MASK: u32 = 0x7ff00000;
```
They will be used by some functions for the future.

File: libseccomp-sys/src/lib.rs [hunk 6]
```diff
@@ -3,22 +3,196 @@
...
+pub const fn SCMP_ACT_ERRNO(x: u16) -> u32 {
+    0x00050000_u32 | x as u32
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:89]: Please change:
```rust
  SCMP_ACT_ERRNO_MASK | x as u32
```

File: libseccomp-sys/src/lib.rs [hunk 7]
```diff
@@ -3,22 +3,196 @@
...
 }
+pub const fn SCMP_ACT_TRACE(x: u16) -> u32 {
+    0x7ff00000_u32 | x as u32
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:92]: Please change:
```rust
SCMP_ACT_TRACE_MASK | x as u32
```

**Review by ManaSugi — DISMISSED** [review]

**Comment by rusty-snake** [comment]
> In addition, we need the tool to update libseccomp-sys for new version, so It's worth leaving tool for the developer.

I don't really see how tool is helpful when updating libseccom-sys to a new version but if you want we can keep it.

As alternative we could replace it with a shell script: 

<details><summary><code>tool.sh</code></summary>

<sub>GitHub tends to mess with tabs&spaces.</sub>

```bash
#!/bin/bash

set -eo pipefail

usage() {
	cat <<-EOM
	USAGE:
	  ./tool.sh [-h] -i <seccomp.h> -o <seccomp.h.rs>
		
	-h  Show this help and exit.
	-i  path to seccomp.h
	-o  path to seccomp.h.rs
	EOM
}

while getopts "hi:o:" opt; do
	case $opt in
		h)
			usage
			exit 0
		;;
		i)
			INPUT="$OPTARG"
		;;
		o)
			OUTPUT="$OPTARG"
		;;
		*)
			exit 2
		;;
	esac
done

if [[ -z "$INPUT" || -z "$OUTPUT" ]]; then
	usage
	exit 2
fi

bindgen \
	--default-enum-style=rust \
	--no-layout-tests \
	--no-doc-comments \
	--allowlist-type="SCMP_.*" \
	--allowlist-type="scmp_.*" \
	--allowlist-type="seccomp_.*" \
	--allowlist-function="seccomp_.*" \
	--allowlist-var=".*SCMP_.*" \
	--output "$OUTPUT"  \
	"$INPUT"
```

</details>

File: libseccomp-sys/src/lib.rs [hunk 8]
```diff
@@ -3,22 +3,199 @@
...
+    pub fn seccomp_export_pfc(ctx: /* FIXME: *const c_void */ scmp_filter_ctx, fd: c_int) -> c_int;
+    pub fn seccomp_export_bpf(ctx: /* FIXME: *const c_void */ scmp_filter_ctx, fd: c_int) -> c_int;
+}
+
+pub const __NR_SCMP_ERROR: c_int = -1;
+pub const __NR_SCMP_UNDEF: c_int = -2;
 
-include!("./libseccomp_bindings.rs");
+// Maybe use https://github.com/seccomp/libseccomp/blob/main/src/syscalls.csv
+// include!("seccomp-syscalls.h.rs");
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:201]: Are theses comments required?

File: libseccomp-sys/src/lib.rs [hunk 9]
```diff
@@ -3,22 +3,199 @@
...
+    pub fn seccomp_api_get() -> c_uint;
+    pub fn seccomp_api_set(level: c_uint) -> c_int;
+    pub fn seccomp_init(def_action: u32) -> scmp_filter_ctx;
+    pub fn seccomp_reset(ctx: scmp_filter_ctx, def_action: u32) -> c_int;
+    pub fn seccomp_release(ctx: scmp_filter_ctx);
+    pub fn seccomp_merge(ctx_dst: scmp_filter_ctx, ctx_src: scmp_filter_ctx) -> c_int;
+    pub fn seccomp_arch_resolve_name(arch_name: *const c_char) -> u32;
+    pub fn seccomp_arch_native() -> u32;
+    pub fn seccomp_arch_exist(
+        ctx: /* FIXME: *const c_void */ scmp_filter_ctx,
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:139]: Do we need to make `scmp_filter_ctx` `const`?

Comment (ManaSugi) [reply libseccomp-sys/src/lib.rs:139]: I agree with we just stick to `*mut c_void`/scmp_filter_ctx` because `scmp_filter_ctx` type is easy to understand.

Comment (ManaSugi) [reply libseccomp-sys/src/lib.rs:139]: Hmm, but it would be better to use `type const_scmp_filter_ctx = *const c_void` to make `libseccomp-sys` more accurate.

File: libseccomp-sys/src/lib.rs [hunk 10]
```diff
@@ -3,22 +3,194 @@
...
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(C)]
+pub struct scmp_arg_cmp {
+    pub arg: c_uint,
+    pub op: scmp_compare,
+    pub datum_a: scmp_datum_t,
+    pub datum_b: scmp_datum_t,
+}
+
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:62]: Could you add `pub const SECCOMP_USER_NOTIF_FLAG_CONTINUE: u32 = 1;`?
I noticed that we need the flag for Seccomp Notification.

Comment (ManaSugi) [reply libseccomp-sys/src/lib.rs:62]: It would be better to generate `libseccomp-sys` with `-allowlist-var=".*SECCOMP_.*"`.

Comment (rusty-snake) [reply libseccomp-sys/src/lib.rs:62]: Do we need any other `linux/seccomp.h` constants?

Comment (rusty-snake) [reply libseccomp-sys/src/lib.rs:62]: I think I just add all the constants from `linux/seccomp.h` as well (except for `SECCOMP_IO*`).

Comment (ManaSugi) [reply libseccomp-sys/src/lib.rs:62]: Yes, could you try to add all the constants from` linux/seccomp.h` (except for SECCOMP_IO*)?

**Comment by ManaSugi** [comment]
Regarding the code, LGTM

**Comment by rusty-snake** [comment]
Squash and rebase.

**Comment by ManaSugi** [comment]
a few minutes ago, I installed [DCO](https://github.com/apps/dco), so please add `Signed-off-by` and push it again.
Sorry for the inconvenience.

**Comment by ManaSugi** [comment]
re-run all jobs

**Comment by ManaSugi** [comment]
>repr(C) on enums can cause UB because we never really know the size of the used integer (for example if libseccomp was >compile with -fshort-enums). However it looks like there is less to nothing that can be done about this (on our end).

I think there is no responsibility about it on our side. So, please cross out.

**Comment by ManaSugi** [comment]
Thank you, please `chmod +x tool.sh` and re-push.
Could you add doc comments ?

**Comment by ManaSugi** [comment]
@rusty-snake Thank you for adding the doc comments. Great work!! LGTM.
Could you remove `tool` in codecov.yml [https://github.com/libseccomp-rs/libseccomp-rs/blob/main/.codecov.yml]?
In addition, please squash the commits into `Manage libseccomp-sys by hand` and `Add tool.sh`.
I think we can land this to main.

**Comment by ManaSugi** [comment]
@rusty-snake  Thank you. Could you confirm the checklist?

File: scripts/tool.sh [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+bindgen \
+	--default-enum-style=rust \
+	--no-layout-tests \
+	--no-doc-comments \
+	--allowlist-type="SCMP_.*" \
+	--allowlist-type="scmp_.*" \
+	--allowlist-type="seccomp_.*" \
+	--allowlist-function="seccomp_.*" \
+	--allowlist-var=".*SCMP_.*" \
```
Comment (ManaSugi) [inline scripts/tool.sh:59]: Could you add `--allowlist-var="SECCOMP_.*`?
Otherwise, we cannot generate `SECCOMP_MODE_DISABLED`, `SECCOMP_USER_NOTIF_FLAG_CONTINUE`, etc.

File: libseccomp-sys/src/lib.rs [hunk 12]
```diff
@@ -3,22 +3,641 @@
...
 #![allow(non_snake_case)]
-#![allow(clippy::all)]
 
-/// Manually support SCMP_ACT_ERRNO and SCMP_ACT_TRACE macro
-/// because the bindgen cannot expand the macros correctly.
+//! Raw FFI bindings for libseccomp library
+
+use std::os::raw::*;
+
+pub const SECCOMP_MODE_DISABLED: u64 = 0;
```
Comment (ManaSugi) [inline libseccomp-sys/src/lib.rs:13]: Why did you change them from `u32` to `u64`?

File: scripts/tool.sh (hunk 11 above)
Comment (rusty-snake) [reply scripts/tool.sh:59]: `#define` constants doesn't have a type, I then usually look at the function where those constants are used which is `prctl` in this case. `prctl` takes them as `arg2` which is an `unsigned long`.

File: libseccomp-sys/src/lib.rs (hunk 12 above)
Comment (rusty-snake) [reply libseccomp-sys/src/lib.rs:13]: `#define` constants doesn't have a type, I then usually look at the function where those constants are used which is `prctl` in this case. `prctl` takes them as `arg2` which is an `unsigned long`.

Comment (ManaSugi) [reply libseccomp-sys/src/lib.rs:13]: Great! I understand. Thank you

**Comment by ManaSugi** [comment]
Could you squash the commits? after squashing, I'll merge this into main.

**Comment by ManaSugi** [comment]
Could you update https://github.com/libseccomp-rs/libseccomp-rs/tree/main/libseccomp-sys/third-party to `2.5.3`?
Or we may not need this。

**Comment by ManaSugi** [comment]
Thank you!

**Review by ManaSugi — APPROVED** [review]
